File: ios/RezumateNative/Services/ReviewPromptTracker.swift
import Foundation

struct ReviewPromptTracker {
    private enum Key {
        static let successfulAnalyses = "rezumate.review.successful-analyses"
        static let successfulExports = "rezumate.review.successful-exports"
        static let lastPromptedVersion = "rezumate.review.last-prompted-version"
    }

    static func recordSuccessfulAnalysis(defaults: UserDefaults = .standard) {
        defaults.set(defaults.integer(forKey: Key.successfulAnalyses) + 1, forKey: Key.successfulAnalyses)
    }

    static func recordSuccessfulExport(defaults: UserDefaults = .standard) {
        defaults.set(defaults.integer(forKey: Key.successfulExports) + 1, forKey: Key.successfulExports)
    }

    static func isEligible(
        marketingVersion: String,
        defaults: UserDefaults = .standard
    ) -> Bool {
        defaults.integer(forKey: Key.successfulAnalyses) >= 2
            && defaults.integer(forKey: Key.successfulExports) >= 1
            && defaults.string(forKey: Key.lastPromptedVersion) != marketingVersion
    }

    static func markPrompted(
        marketingVersion: String,
        defaults: UserDefaults = .standard
    ) {
        defaults.set(marketingVersion, forKey: Key.lastPromptedVersion)
    }
}

File: android/app/src/main/java/com/aftaab/rezumate/domain/ReviewPromptTracker.kt
package com.aftaab.rezumate.domain

interface ReviewPromptStore {
    suspend fun successfulAnalyses(): Int
    suspend fun successfulExports(): Int
    suspend fun lastPromptedVersion(): String?
    suspend fun setSuccessfulAnalyses(value: Int)
    suspend fun setSuccessfulExports(value: Int)
    suspend fun setLastPromptedVersion(value: String)
}

class InMemoryReviewPromptStore : ReviewPromptStore {
    private var analyses = 0
    private var exports = 0
    private var lastPrompted: String? = null

    override suspend fun successfulAnalyses(): Int = analyses
    override suspend fun successfulExports(): Int = exports
    override suspend fun lastPromptedVersion(): String? = lastPrompted
    override suspend fun setSuccessfulAnalyses(value: Int) {
        analyses = value
    }
    override suspend fun setSuccessfulExports(value: Int) {
        exports = value
    }
    override suspend fun setLastPromptedVersion(value: String) {
        lastPrompted = value
    }
}

class ReviewPromptTracker(
    private val store: ReviewPromptStore,
) {
    suspend fun recordSuccessfulAnalysis() {
        store.setSuccessfulAnalyses(store.successfulAnalyses() + 1)
    }

    suspend fun recordSuccessfulExport() {
        store.setSuccessfulExports(store.successfulExports() + 1)
    }

    suspend fun isEligible(marketingVersion: String): Boolean =
        store.successfulAnalyses() >= 2 &&
            store.successfulExports() >= 1 &&
            store.lastPromptedVersion() != marketingVersion

    suspend fun markPrompted(marketingVersion: String) {
        store.setLastPromptedVersion(marketingVersion)
    }
}
